packages/puppeteer-core/src/injected/PSelectorParser.ts

```diff
@@ -51,69 +51,74 @@ const unquote = (text: string): string => {
+    const tokens = tokenize(selector);
+    if (tokens.length === 0) {
+      return [[], isPureCSS];
+    }
+    let compoundSelector: CompoundPSelector = [];
+    let complexSelector: ComplexPSelector = [compoundSelector];
+    const selectors: ComplexPSelectorList = [complexSelector];
+    const storage: Token[] = [];
+    for (const token of tokens) {
+      switch (token.type) {
+        case 'combinator':
+          switch (token.content) {
+            case PCombinator.Descendent:
+              isPureCSS = false;
+              if (storage.length) {
+                compoundSelector.push(stringify(storage));
+                storage.splice(0);
+              }
+              compoundSelector = [];
+              complexSelector.push(PCombinator.Descendent);
+              complexSelector.push(compoundSelector);
+              continue;
+            case PCombinator.Child:
+              isPureCSS = false;
+              if (storage.length) {
+                compoundSelector.push(stringify(storage));
+                storage.splice(0);
+              }
+              compoundSelector = [];
+              complexSelector.push(PCombinator.Child);
+              complexSelector.push(compoundSelector);
+              continue;
+          }
           break;
-        }
-        isPureCSS = false;
-        if (storage.length) {
-          compoundSelector.push(stringify(storage));
-          storage.splice(0);
-        }
-        compoundSelector.push({
-          name: token.name.slice(3),
-          value: unquote(token.argument ?? ''),
-        });
-        continue;
-      case 'comma':
-        if (storage.length) {
-          compoundSelector.push(stringify(storage));
-          storage.splice(0);
-        }
-        compoundSelector = [];
-        complexSelector = [compoundSelector];
-        selectors.push(complexSelector);
-        continue;
+        case 'pseudo-element':
+          if (!token.name.startsWith('-p-')) {
+            break;
+          }
+          isPureCSS = false;
+          if (storage.length) {
+            compoundSelector.push(stringify(storage));
+            storage.splice(0);
+          }
+          compoundSelector.push({
+            name: token.name.slice(3),
+            value: unquote(token.argument ?? ''),
+          });
+          continue;
+        case 'comma':
+          if (storage.length) {
+            compoundSelector.push(stringify(storage));
+            storage.splice(0);
+          }
+          compoundSelector = [];
+          complexSelector = [compoundSelector];
+          selectors.push(complexSelector);
+          continue;
+      }
+      storage.push(token);
     }
-    storage.push(token);
-  }
-  if (storage.length) {
-    compoundSelector.push(stringify(storage));
+    if (storage.length) {
+      compoundSelector.push(stringify(storage));
+    }
+    return [selectors, isPureCSS];
+  } catch {
+    // In case parsing breaks, we fallback to CSS.
+    return [[], true];
```

Comment:
Let's do the try/catch on the consumer side. I don't think p-selectors should be concerned with fallbacks and know about CSS.

Reply:
SGTM